generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java

```diff
@@ -940,47 +940,62 @@ private void generateClassWriter(String className, StructSpec struct,
                         forVersions(field.versions(), field.taggedVersions().intersect(field.versions())).
                         allowMembershipCheckAlwaysFalse(false).
                         ifMember(presentAndTaggedVersions -> {
-                            generateNonDefaultValueCheck(field);
-                            buffer.incrementIndent();
-                            buffer.printf("_writable.writeUnsignedVarint(%d);%n", field.tag().get());
-                            if (field.type().isString()) {
-                                IsNullConditional.forName(field.camelCaseName()).
-                                    nullableVersions(field.nullableVersions()).
-                                    possibleVersions(presentAndTaggedVersions).
-                                    ifNull(() -> {
-                                        buffer.printf("_writable.writeUnsignedVarint(0);%n");
-                                    }).
-                                    ifNotNull(() -> {
+                            IsNullConditional cond = IsNullConditional.forName(field.camelCaseName()).
+                                nullableVersions(field.nullableVersions()).
+                                possibleVersions(presentAndTaggedVersions).
+                                alwaysEmitBlockScope(true).
+                                ifNotNull(() -> {
+                                    if (!field.defaultString().equals("null")) {
+                                        generateNonDefaultValueCheck(field, Versions.NONE);
+                                        buffer.incrementIndent();
+                                    }
+                                    buffer.printf("_writable.writeUnsignedVarint(%d);%n", field.tag().get());
+                                    if (field.type().isString()) {
                                         buffer.printf("byte[] _stringBytes = _cache.getSerializedValue(this.%s);%n",
                                             field.camelCaseName());
-                                        buffer.printf("_writable.writeUnsignedVarint(_stringBytes.length);%n");
+                                        headerGenerator.addImport(MessageGenerator.BYTE_UTILS_CLASS);
+                                        buffer.printf("_writable.writeUnsignedVarint(_stringBytes.length + " +
```

Comment:
So the reason we have to write the size twice here is because we treat all tagged fields as structs. Is that right? So we could convert "myNullableString" into a struct which has a nullable string as its first field and then is followed by additional fields. Assuming that is right, I'm wondering how it works in terms of versioning. Do tagged fields have prefix rules? 

By the way, what is the reason for the '+ 1'?

Reply:
It's not really that they're all treated as structs, but just that they all have to be treated as opaque length-delimited blobs so that clients that don't know the type of the tagged field can skip over it.  The length field is ignored by clients that know the field type (for example, if you know that you're looking for an int, you can just read 4 bytes).  But if you don't know about the field, you also don't know about the field's type.

I guess in theory we could support inline tagged structs if we wanted to later on.  Since everything is just length-prefixed, we can add arbitrary types that nobody has thought of yet in a backwards-compatible fashion.  Right now, they're not supported, though.

The way compatibility works is basically "you can never change the type of a tagged field."  The constraint isn't as bad as it sounds, since you can always create a new tagged field if you really need something of a different type.

Reply:
Yes, I understand the need to have a length at the tagged field level. Mainly I was thinking about the redundancy when the type of the tagged field is known. It's a little annoying to serialize the string length twice, for example. Anyway, thanks for explaining.